src/constraint.rs

```diff
@@ -703,3 +715,62 @@ impl ConstraintValidator for FieldsConstraint {
         Ok(())
     }
 }
+
+/// Implements an `contains` constraint of Ion Schema
+/// [contains]: https://amzn.github.io/ion-schema/docs/spec.html#contains
+#[derive(Debug, Clone, PartialEq)]
+pub struct ContainsConstraint {
+    // TODO: convert this into a HashSet once we have an implementation of Hash for OwnedElement in ion-rust
+    // Reference ion-rust issue: https://github.com/amzn/ion-rust/issues/220
+    values: Vec<OwnedElement>,
+}
+
+impl ContainsConstraint {
+    pub fn new(values: Vec<OwnedElement>) -> Self {
+        Self { values }
+    }
+}
+
+impl ConstraintValidator for ContainsConstraint {
+    fn validate(&self, value: &OwnedElement, type_store: &TypeStore) -> ValidationResult {
+        // get the expected values from the contains constraint
+        let mut expected_values = self.values.clone();
+
+        // Check for null sequence
+        if value.is_null() {
+            return Err(Violation::new(
+                "contains",
+                ViolationCode::TypeMismatched,
+                &format!("Null sequence not allowed for contains constraint"),
+            ));
+        }
+
+        let values: Vec<&OwnedElement> = match value.as_sequence() {
+            None => {
+                return Err(Violation::new(
+                    "contains",
+                    ViolationCode::TypeMismatched,
+                    &format!("expected list found {}", value.ion_type()),
+                ));
+            }
+            Some(ion_sequence) => ion_sequence.iter().collect(),
+        };
+
+        for value in values {
+            if let Some(pos) = expected_values.iter().position(|v| v == value) {
+                expected_values.remove(pos);
+            }
+        }
```

Comment:
Rather than cloning and collecting all of the collections, I think we can just do this:

```
// If any of our expected values are not in the Ion sequence, this is false
expected_values
    .iter()
    .all(|ev| ion_sequence.iter().contains(ev))
```

It's `O(n^2)`, but you've already got a TODO to address that.

https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.all
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.contains

Reply:
Using `find` instead of `contains` to be able to use an iterator and to collect values to a vector.

https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find